test/serverExplorer.test.ts

```diff
@@ -0,0 +1,193 @@
+import * as chai from 'chai';
+import * as sinon from 'sinon';
+import * as sinonChai from 'sinon-chai';
+import { ServersViewTreeDataProvider } from '../src/serverExplorer';
+import { RSPClient, Protocol } from 'rsp-client';
+import { EventEmitter, window } from 'vscode';
+
+const expect = chai.expect;
+chai.use(sinonChai);
+
+suite('Server explorer', () => {
+
+    let sandbox: sinon.SinonSandbox;
+    const clientStub: sinon.SinonStubbedInstance<RSPClient> = new RSPClient('somehost', 8080);
+    let serverExplorer: ServersViewTreeDataProvider;
+
+    setup(() => {
+        serverExplorer = new ServersViewTreeDataProvider(clientStub);
+        sandbox = sinon.createSandbox();
+    });
+
+    teardown(() => {
+        sandbox.restore();
+    });
+
+    const serverType: Protocol.ServerType = {
+        description: 'a type',
+        id: 'type',
+        visibleName: 'the type'
+    };
+
+    const serverHandle: Protocol.ServerHandle = {
+        id: 'id',
+        type: serverType
+    };
+
+    const stateChange: Protocol.ServerStateChange = {
+        server: serverHandle,
+        state: 1
+    };
+
+    const ProcessOutput: Protocol.ServerProcessOutput = {
+        processId: 'process id',
+        server: serverHandle,
+        streamType: 0,
+        text: 'the type'
+    };
+
+    const findServerBeans = {
+        length: 1,
+        fullVersion: 'version',
+        location: 'path',
+        name: 'EAP',
+        serverAdapterTypeId: 'org.jboss',
+        specificType: 'EAP',
+        typeCategory: 'EAP',
+        version: '7.1'
+    };
+
+    const status = {
+        code: 0,
+        message: 'ok',
+        pluginId: 'unknown',
+        severity: 0
+    };
+
+    test('InsertServer call should add server to tree data model', () => {
+        const refreshStub = sandbox.stub(serverExplorer, 'refresh').returns(null);
+        const getChildren = serverExplorer.getChildren(null);
+        serverExplorer.insertServer(serverHandle);
+        expect(refreshStub).calledOnce;
+        expect(getChildren).deep.equals(undefined);
+    });
+
+    test('removeServer call should remove server from tree data model', () => {
+        const disposeStub = sandbox.stub(serverExplorer.serverOutputChannels, 'get').returns({
+            clear: () => {},
+            dispose: () => {}
+        });
+        const getChildren = serverExplorer.getChildren(null);
+        sandbox.stub(serverExplorer, 'refresh');
+        serverExplorer.removeServer(serverHandle);
+        expect(disposeStub).calledOnce;
+        expect(getChildren).deep.equals(undefined);
+    });
+
+    test('serverExplorer.updateServer call should update server state to received in state change event', () => {
+        const serviceStub = sandbox.stub(serverExplorer.servers, 'get').returns({
```

Comment:
I am not following what is going on here. You should first setup the condition of the test and initialize model the way it would look like there is a server in the model in specific state. Let say the state is 'stopped'. Then you inject the even that would inform the model about server changing its state to 'starting'. Then you verify that specific methods were called and server stored in the model changed its state to exactly that one you injected in the beginning of the test.